runtime/neumann/src/fee_handler.rs

```diff
@@ -0,0 +1,285 @@
+};
+use sp_std::marker::PhantomData;
+
+#[derive(Debug)]
+pub struct FeeInformation {
+	token_id: TokenId,
+	xcm_data: Option<XcmCurrencyData>,
+	asset_metadata: Option<AssetMetadata<Balance, CustomMetadata>>,
+}
+pub trait CallParser<Call> {
+	fn fee_information(call: &Call) -> FeeInformation;
+}
+pub struct FeeCallParser;
+impl CallParser<Call> for FeeCallParser {
+	fn fee_information(c: &Call) -> FeeInformation {
+		if let Call::AutomationTime(pallet_automation_time::Call::schedule_xcmp_task {
+			para_id,
+			currency_id,
+			..
+		}) = c.clone()
+		{
+			let xcm_data = XcmpHandler::get_xcm_chain_data(u32::from(para_id), currency_id);
+			let asset_metadata = AssetRegistry::metadata(currency_id);
+
+			FeeInformation { token_id: currency_id, xcm_data, asset_metadata }
+		} else {
+			FeeInformation { token_id: NATIVE_TOKEN_ID, xcm_data: None, asset_metadata: None }
+		}
+	}
+}
+
+pub struct NativeFeeProcessor<T, C>(PhantomData<(T, C)>);
+impl<T, C> NativeFeeProcessor<T, C>
+where
+	T: pallet_transaction_payment::Config,
+	C: Currency<<T as frame_system::Config>::AccountId>,
+{
+	fn withdraw_fee(
+		who: &T::AccountId,
+		fee: C::Balance,
+		tip: C::Balance,
+	) -> Result<Option<NegativeImbalanceOf<C, T>>, TransactionValidityError> {
+		let withdraw_reason = if tip.is_zero() {
+			WithdrawReasons::TRANSACTION_PAYMENT
+		} else {
+			WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+		};
+
+		match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+			Ok(imbalance) => Ok(Some(imbalance)),
+			Err(_) => Err(InvalidTransaction::Payment.into()),
+		}
+	}
+}
+
+pub struct ForeignFeeProcessor<T, MC, TA>(PhantomData<(T, MC, TA)>);
+impl<T, MC, TA> ForeignFeeProcessor<T, MC, TA>
+where
+	T: pallet_transaction_payment::Config,
+	MC: MultiCurrency<<T as frame_system::Config>::AccountId>,
+	MC::CurrencyId: From<TokenId>,
+	TA: Get<T::AccountId>,
+{
+	fn withdraw_fee(
+		who: &T::AccountId,
+		fee: MC::Balance,
+		call_information: FeeInformation,
+	) -> Result<(), TransactionValidityError> {
+		if call_information.xcm_data.is_none() || call_information.asset_metadata.is_none() {
+			return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
+		}
+
+		let currency_id = call_information.token_id.into();
+		let foreign_fee: MC::Balance = call_information
+			.asset_metadata
+			.ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?
+			.convert_fee_into_foreign(fee.saturated_into())
+			.ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?
+			.saturated_into();
+
+		// orml_tokens doesn't provide withdraw that allows setting existence so prevent
+		// withdraw if they don't have enough to avoid reaping
+		MC::ensure_can_withdraw(
+			currency_id,
+			who,
+			foreign_fee.saturating_add(MC::minimum_balance(currency_id)),
+		)
+		.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+		MC::withdraw(currency_id, who, foreign_fee)
```

Comment:
remind me why we aren't just using a transfer?

Reply:
`transfer` still doesn't allow setting existence.  I think we decided they were the same and it didn't matter.  I'm happy with whichever unless @nvengal remembers something we don't.

Reply:
My opinion, loosely held, is that it's still best to stick with the pattern of `withdraw` then `deposit` for handling fees. The pattern for fees seems to be withdraw and then handle or burn imbalance. I think there's value in sticking to the general pattern. But, again, loosely held. I'm having trouble figuring out if `transfer` has any read/write savings associated.